TestScript datatypes/timestamp-with-time-zone.sql fails if TZ=Europe/Berlin

When running
`./build.sh testTravis`
with TZ=Europe/Berlin 
against the latest source the test fails with:
`22:33:43 00:10.597 org.h2.test.scripts.TestScript Running commands in datatypes/timestamp-with-time-zone.sql
ERROR: Unexpected General error SELECT CAST (CAST (TIMESTAMP '1000000000-12-31 23:59:59.999999999' AS TIMESTAMP(9) WITH TIME ZONE) AS TIMESTAMP(9)) org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.time.DateTimeException: Invalid value for Year (valid values -999999999 - 999999999): 1000000000"; SQL statement:
SELECT CAST (CAST (TIMESTAMP '1000000000-12-31 23:59:59.999999999' AS TIMESTAMP(9) WITH TIME ZONE) AS TIMESTAMP(9)) [50000-200] ------------------------------
org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.time.DateTimeException: Invalid value for Year (valid values -999999999 - 999999999): 1000000000"; SQL statement:
SELECT CAST (CAST (TIMESTAMP '1000000000-12-31 23:59:59.999999999' AS TIMESTAMP(9) WITH TIME ZONE) AS TIMESTAMP(9)) [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:500)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:424)
	at org.h2.message.DbException.get(DbException.java:189)
	at org.h2.message.DbException.convert(DbException.java:342)
	at org.h2.command.Command.executeQuery(Command.java:210)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:230)
	at org.h2.test.scripts.TestScript.processStatement(TestScript.java:528)
	at org.h2.test.scripts.TestScript.process(TestScript.java:445)
	at org.h2.test.scripts.TestScript.testFile(TestScript.java:380)
	at org.h2.test.scripts.TestScript.testScript(TestScript.java:266)
	at org.h2.test.scripts.TestScript.test(TestScript.java:159)
	at org.h2.test.TestBase.runTest(TestBase.java:141)
	at org.h2.test.TestAll.addTest(TestAll.java:1005)
	at org.h2.test.TestAll.test(TestAll.java:722)
	at org.h2.test.TestAll.runTests(TestAll.java:612)
	at org.h2.test.TestAll.testAll(TestAll.java:555)
	at org.h2.test.TestAll.run(TestAll.java:505)
	at org.h2.test.TestAll.main(TestAll.java:447)
Caused by: java.time.DateTimeException: Invalid value for Year (valid values -999999999 - 999999999): 1000000000
	at java.time.temporal.ValueRange.checkValidIntValue(ValueRange.java:330)
	at java.time.temporal.ChronoField.checkValidIntValue(ChronoField.java:722)
	at java.time.LocalDate.ofEpochDay(LocalDate.java:341)
	at java.time.zone.ZoneRules.findYear(ZoneRules.java:928)
	at java.time.zone.ZoneRules.getOffset(ZoneRules.java:487)
	at org.h2.util.TimeZoneProvider$WithTimeZone.getTimeZoneOffsetUTC(TimeZoneProvider.java:367)
	at org.h2.value.Value.convertToTimestamp(Value.java:1207)
	at org.h2.value.Value.convertTo(Value.java:787)
	at org.h2.value.Value.convertTo(Value.java:739)
	at org.h2.value.TypeInfo.cast(TypeInfo.java:501)
	at org.h2.expression.function.Function.getSimpleValue(Function.java:976)
	at org.h2.expression.function.Function.getValueWithArgs(Function.java:1303)
	at org.h2.expression.function.Function.getValue(Function.java:679)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1774)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:101)
	at org.h2.result.LazyResult.next(LazyResult.java:60)
	at org.h2.command.dml.Select.queryFlat(Select.java:720)
	at org.h2.command.dml.Select.queryWithoutCache(Select.java:826)
	at org.h2.command.dml.Query.queryWithoutCacheLazyCheck(Query.java:190)
	at org.h2.command.dml.Query.query(Query.java:475)
	at org.h2.command.dml.Query.query(Query.java:438)
	at org.h2.command.CommandContainer.query(CommandContainer.java:262)
	at org.h2.command.Command.executeQuery(Command.java:189)
	... 13 more
ERROR: org/h2/test/scripts/datatypes/timestamp-with-time-zone.sql
line: 142
exp: >> 1000000000-12-31 23:59:59.999999999
got: > exception GENERAL_ERROR_1
`

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2364: TestScript datatypes/timestamp-with-time-zone.sql fails if TZ=Europe/Berlin
+</li>
 <li>Issue #2359: Complete implementation of generated columns
 </li>
 <li>PR #2361: Fix unused result

--- a/h2/src/main/org/h2/util/TimeZoneProvider.java
+++ b/h2/src/main/org/h2/util/TimeZoneProvider.java
@@ -321,10 +321,6 @@ public abstract class TimeZoneProvider {
          */
         static final long SECONDS_PER_YEAR = SECONDS_PER_PERIOD / 400;
 
-        private static final long EPOCH_SECONDS_HIGH = 31556889864403199L;
-
-        private static final long EPOCH_SECONDS_LOW = -31557014167219200L;
-
         private static volatile DateTimeFormatter TIME_ZONE_FORMATTER;
 
         private final ZoneId zoneId;
@@ -353,15 +349,19 @@ public abstract class TimeZoneProvider {
         public int getTimeZoneOffsetUTC(long epochSeconds) {
             /*
              * Construct an Instant with EPOCH seconds within the range
-             * -31,557,014,167,219,200..31,556,889,864,403,199
-             * (-1000000000-01-01T00:00Z..1000000000-12-31T23:59:59.999999999Z).
-             * Too large and too small EPOCH seconds are replaced with EPOCH
-             * seconds within the range using the 400 years period of the
-             * Gregorian calendar.
+             * -31,557,014,135,532,000..31,556,889,832,715,999
+             * (-999999999-01-01T00:00-18:00..
+             * +999999999-12-31T23:59:59.999999999+18:00). Too large and too
+             * small EPOCH seconds are replaced with EPOCH seconds within the
+             * range using the 400 years period of the Gregorian calendar.
+             *
+             * H2 has slightly wider range of EPOCH seconds than Instant, and
+             * ZoneRules.getOffset(Instant) does not support all Instant values
+             * in all time zones.
              */
-            if (epochSeconds > EPOCH_SECONDS_HIGH) {
+            if (epochSeconds > 31_556_889_832_715_999L) {
                 epochSeconds -= SECONDS_PER_PERIOD;
-            } else if (epochSeconds < EPOCH_SECONDS_LOW) {
+            } else if (epochSeconds < -31_557_014_135_532_000L) {
                 epochSeconds += SECONDS_PER_PERIOD;
             }
             return zoneId.getRules().getOffset(Instant.ofEpochSecond(epochSeconds)).getTotalSeconds();

